utils/arch.h

```diff
@@ -24,6 +25,10 @@ static inline enum uftrace_cpu_arch host_cpu_arch(void)
 	return UFT_CPU_AARCH64;
 #elif defined(__i386__)
 	return UFT_CPU_I386;
+#elif defined(__riscv)
+#if __riscv_xlen == 64
+	return UFT_CPU_RISCV64;
+#endif
```

Comment:
It'd make it undefined if `__riscv` is defined but `__riscv_xlen` is not 64.  You need to make it in a single condition like

```c
#elif defined(__riscv) && __riscv_xlen == 64
        return UFT_CPU_RISCV64;
#else
```

Reply:
The `__riscv` macro seems to point to all RISC-V architectures (riscv32, riscv64, riscv128).

So I thought it would be a good idea to separate them, considering that `riscv32` should be added in the future.

Would it still be better to configure the if statement with a single condition, as you suggested?

Reply:
According to the link below, older GCCs are using `__riscv__` instead of `__riscv`.

Therefore, we will add a condition that also checks for `__riscv__`.
<br>

Ref. https://github.com/riscv-non-isa/riscv-toolchain-conventions#cc-preprocessor-definitions

Reply:
> Would it still be better to configure the if statement with a single condition, as you suggested?

It should return a sane value for any configuration.  But it seems it doesn't return anything for `riscv32`.

> older GCCs are using __riscv__ instead of __riscv.

How old are those GCCs?  I'm not sure if we really care about them.

Reply:
> But it seems it doesn't return anything for riscv32.

Hmm...  I didn't understand this part yet.

But I'm guessing that in the future, if we say we support both riscv64 and riscv32, the code with the above feedback would look like this, what do you think?
```
#elif defined(__riscv) && __riscv_xlen == 64
        return UFT_CPU_RISCV64;
#elif defined(__riscv) && __riscv_xlen == 32
        return UFT_CPU_RISCV32;
#else
```
<br>

------

Ref. the code in the official GCC Mirror Repository is implemented as follows.

in gcc/config/riscv/riscv.h
```
#define TARGET_64BIT (__riscv_xlen == 64)

......

#define UNITS_PER_WORD (TARGET_64BIT ? 8 : 4)
```

in gcc/config/riscv/riscv-c.cc
```
builtin_define_with_int_value ("__riscv_xlen", UNITS_PER_WORD * 8);
```

Reply:
> How old are those GCCs? I'm not sure if we really care about them.

To conclude, I think we shouldn't bother with `__riscv__` as @namhyung said.

This is because the official GCC Mirror Repository first supported RISC-V in `gcc-7.1.0`, and that version only supports the `__riscv` macro.
  - https://github.com/gcc-mirror/gcc/blob/releases/gcc-7.1.0/gcc/config/riscv/riscv-c.c
<br>

------
However, the reference link I referenced above says that older GCCs use `__riscv__`.

However, there is very little information about `__riscv__` and it was hard to find.

Therefore, my conclusion is that older GCC is riscv-gcc, which seems to have been maintained on its own before RISC-V was integrated into official GCC, and the relevant commit is below. 

riscv-collab/riscv-gcc@b6fc2268f59

Reply:
>>   But it seems it doesn't return anything for riscv32.

> Hmm... I didn't understand this part yet.

My understanding is that current code is doing this:

```c

static inline enum uftrace_cpu_arch host_cpu_arch(void)
{
#if defined(__x86_64__)
	return UFT_CPU_X86_64;
#elif defined(__arm__)
	return UFT_CPU_ARM;
#elif defined(__aarch64__)
	return UFT_CPU_AARCH64;
#elif defined(__i386__)
	return UFT_CPU_I386;
#elif defined(__riscv)
# if __riscv_xlen == 64
	return UFT_CPU_RISCV64;
# endif
#else
	return UFT_CPU_NONE;
#endif
}
```

If `__riscv` is defined but `__riscv_xlen` is 32 (or other than 64) there's no return statement for it.  So yeah, you need to check the xlen size properly and handle the else case.

Reply:
Ah! I got it now.


If the system is `riscv32`, it satisfies `#elif defined(__riscv)` to get into that `#elif` statement.
But it doesn't return UFT_CPU_NONE because it doesn't satisfy `#if __riscv_xlen == 64`.

I will apply that part as well, thanks.